app/Platform/Services/TicketService.php

```diff
@@ -0,0 +1,96 @@
+
+class TicketService
+{
+    public int $unlocksSinceReported = 0;
+    public array $openTicketLinks = [];
+    public array $closedTicketLinks = [];
+    public string $contactReporterUrl = '';
+    public ?PlayerAchievement $existingUnlock = null;
+    public ?PlayerGameActivityService $activity = null;
+    public ?UserAgentService $userAgentService = null;
+    public string $ticketNotes = '';
+
+    public function load(Ticket $ticket): void
+    {
+        $msgTitle = rawurlencode("Bug Report ({$ticket->achievement->game->Title})");
+        if ($ticket->reporter) {
+            $msgPayload = "Hi [user={$ticket->reporter->User}], I'm contacting you about [ticket={$ticket->ID}]";
+            $msgPayload = rawurlencode($msgPayload);
+            $this->contactReporterUrl = route('message.create') . "?to={$ticket->reporter->User}&subject=$msgTitle&message=$msgPayload";
+
+            $this->existingUnlock = PlayerAchievement::where('user_id', $ticket->reporter->id)
+                ->where('achievement_id', $ticket->achievement->id)
+                ->first();
+
+            $this->openTicketLinks = [];
+            $this->closedTicketLinks = [];
+            $achievementTickets = Ticket::where('AchievementID', $ticket->achievement->id);
+            foreach ($achievementTickets->get() as $otherTicket) {
+                if ($otherTicket->ID !== $ticket->ID) {
+                    $url = '<a href="' . route('ticket.show', ['ticket' => $otherTicket]) . '">' . $otherTicket->ID . '</a>';
+                    if (TicketState::isOpen($otherTicket->ReportState)) {
+                        $this->openTicketLinks[] = $url;
+                    } else {
+                        $this->closedTicketLinks[] = $url;
+                    }
+                }
+            }
+        }
+
+        $this->unlocksSinceReported = 0;
+        if (TicketState::isOpen($ticket->ReportState)) {
+            $this->unlocksSinceReported = PlayerAchievement::where('achievement_id', $ticket->achievement->id)
+                ->where(function ($query) use ($ticket) {
+                    $query->where('unlocked_at', '>', $ticket->ReportedAt)
+                        ->orWhere('unlocked_hardcore_at', '>', $ticket->ReportedAt);
+                })->count();
+        }
+
+        $this->ticketNotes = nl2br($ticket->ReportNotes);
+        foreach ($ticket->achievement->game->hashes as $hash) {
+            if (stripos($this->ticketNotes, $hash->md5) !== false) {
+                $replacement = '<a href="/linkedhashes.php?g=' . $ticket->achievement->game->id . '" title="' .
+                    attributeEscape($hash->name) . '">' . $hash->md5 . '</a>';
+                $this->ticketNotes = str_ireplace($hash->md5, $replacement, $this->ticketNotes);
+            }
+        }
+    }
+
+    public function buildHistory(Ticket $ticket, User $user): void
+    {
+        $this->activity = new PlayerGameActivityService();
+        $this->userAgentService = new UserAgentService();
```

Comment:
nit: These should probably be instantiated via DI in `__construct()`:

```php
public function __construct(
  protected PlayerGameActivityService $activity,
  protected UserAgentService $userAgentService,
) {
}
```

These will be spun up and torn down only during the lifetime of the current request.

Reply:
done